src/python/pants/testutil/rule_runner.py

```diff
@@ -216,23 +217,28 @@ def run_goal_rule(
         console.flush()
         return GoalRuleResult(exit_code, stdout.getvalue(), stderr.getvalue())
 
-    def set_options(self, args: Iterable[str], *, env: Mapping[str, str] | None = None) -> None:
+    def set_options(
+        self, args: Iterable[str], *, env: Mapping[str, str | None] | None = None
+    ) -> None:
         """Update the engine session with new options and/or environment variables.
 
         The environment variables will be used to set the `PantsEnvironment`, which is the
         environment variables captured by the parent Pants process. Some rules use this to be able
         to read arbitrary env vars. Any options that start with `PANTS_` will also be used to set
-        options.
+        options. Giving an environment variable a value of `None` will cause it to be propagated from
+        the test runner's environment (os.environ).
 
         This will override any previously configured values.
         """
+        env = {k: (v if v is not None else os.getenv(k, "")) for k, v in env.items()} if env else {}
```

Comment:
@Eric-Arellano : Interested in your thoughts around this API (check out some of the call sites for examples of how this will be used).

The gist is that we have so far been leaking through (some) environment variables, and rather than adding those variables as defaults, I think we should make them explicit at the rule callsites. Based on the CI run, I'll need to update ~20 test files to explicitly propagate two or three variables from the test process to the `RuleRunner`.

Reply:
I really like that that's explicit now. 

Using `None` reads a little weird at call sites. My other thought is to use a sentinel value like `{"PATH": INHERIT}`. It's not as ergonomic to need to import a value, but it looks like this isn't super common to need to do.

--

Either design, would be good to break this out as a precursor.

Reply:
Extracted to #11641.